onnxruntime/core/optimizer/matmul_add_fusion.cc

```diff
@@ -69,25 +74,32 @@ Status MatMulAddFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       continue;
     }
 
-    auto matmul_output_name = matmul_node.OutputDefs()[0]->Name();
+    const auto& matmul_output = *matmul_node.OutputDefs()[0];
+
+    auto matmul_output_name = matmul_output.Name();
     auto gemm_input_defs = matmul_input_defs;
     if (matmul_output_name == add_input_defs[0]->Name()) {
       // matmul output as Add_A, should use Add_B as input C for gemm
-      // Gemm only support unidirectional broadcast on C
-      if (add_input_defs[1]->Shape()->dim_size() > 2) {
-        continue;
-      }
       gemm_input_defs.push_back(add_input_defs[1]);
     } else {
       // matmul output as Add_B, should use Add_A as input C for gemm
-      // Gemm only support unidirectional broadcast on C
-      if (add_input_defs[0]->Shape()->dim_size() > 2) {
-        continue;
-      }
       gemm_input_defs.push_back(add_input_defs[0]);
     }
 
-    if (!graph.GetNodeOutputsInGraphOutputs(node).empty()) {
+    // valid bias_shapes are (N) or (1, N) or (M, 1) or (M, N) as
+    // GEMM only supports unidirectional broadcast on the bias input C
+    const auto& bias_shape = *gemm_input_defs.back()->Shape();
+    const auto& M = matmul_output.Shape()->dim()[0];
+    const auto& N = matmul_output.Shape()->dim()[1];
+    auto dim_has_value_1 = [](const TensorShapeProto_Dimension& dim) {
+      return dim.has_dim_value() && dim.dim_value() == 1;
+    };
+
+    bool valid = ((bias_shape.dim_size() == 1 && bias_shape.dim()[0] == N) ||
+                  (bias_shape.dim_size() == 2 && dim_has_value_1(bias_shape.dim()[0]) && bias_shape.dim()[1] == N) ||
+                  (bias_shape.dim_size() == 2 && bias_shape.dim()[0] == M &&
+                   (dim_has_value_1(bias_shape.dim()[1]) || bias_shape.dim()[1] == N)));
```

Comment:
Hi, I believe dim_size() can be > 2, since the output is either a 2-D matrix, or a N-D stacks of matrices, where the last 2 dims are  M, N.

So output shape can be  Z = (I, J, K, ...,M, N).
And legal bias_shape values would be  any rightmost substring of Z, with corresponding entries equal to Z or else 1.

e.g. 
(1, M , N)
(1, 1, M, N)
(1, K, 1, M, 1)

Reply:
This applies doesn't it? https://github.com/onnx/onnx/blob/master/docs/Broadcasting.md#unidirectional-broadcasting

Reply:
FWIW this logic is copied from the test in GemmHelper that is used by the GEMM op, so any check here aligns with how ORT GEMM currently works (rightly or wrongly).

Reply:
It looks like the GemmHelper checks will work for 2D matrices -- but probably give incorrect results for general N-D . The link you pasted  above looks correct to me however, since it's got higher dimensional examples and description.

Not sure why the GemmHelper is different from the documentation -- there might be some historical reason for it (which may or may not apply today).

One option is to  update the code/logic in both locations (to make everything work for N-D). 

Another option is to  keep the current 2D logic, and simply not do a op-fusion if shape > 2D, since we won't know if the shapes are truly compatible.,

Reply:
Based on line 72,  it seems like only 2D is supported -- in which case this logic looks Ok. 

Signing off.